plugins/plugin_utils/_reboot.py

```diff
@@ -0,0 +1,469 @@
+    return stdout
+
+
+def _perform_reboot(task_action, connection, reboot_command, handle_abort=True):
+    # type: (str, ConnectionBase, str, bool) -> None
+    """Runs the reboot command"""
+    display.vv("%s: rebooting server..." % task_action)
+
+    stdout = stderr = None
+    try:
+        rc, stdout, stderr = _execute_command(task_action, connection, reboot_command)
+
+    except AnsibleConnectionFailure as e:
+        # If the connection is closed too quickly due to the system being shutdown, carry on
+        display.vvv('%s: AnsibleConnectionFailure caught and handled: %s' % (task_action, to_text(e)))
+        rc = 0
+
+    if stdout:
+        try:
+            reboot_result = json.loads(stdout)
+        except getattr(json.decoder, 'JSONDecodeError', ValueError):
+            # While the reboot command should output json it may have failed for some other reason. We continue
+            # reporting with that output instead
+            pass
+        else:
+            stdout = reboot_result.get('stdout', stdout)
+            stderr = reboot_result.get('stderr', stderr)
+            rc = int(reboot_result.get('rc', rc))
+
+    # Test for "A system shutdown has already been scheduled. (1190)" and handle it gracefully
+    if handle_abort and (rc == 1190 or (rc != 0 and "(1190)" in stderr)):
+        display.warning('A scheduled reboot was pre-empted by Ansible.')
+
+        # Try to abort (this may fail if it was already aborted)
+        rc, stdout, stderr = _execute_command(task_action, connection, 'shutdown.exe /a')
+        display.vvv("%s: result from trying to abort existing shutdown - rc: %s, stdout: %s, stderr: %s"
+                    % (task_action, rc, stdout, stderr))
+
+        return _perform_reboot(task_action, connection, reboot_command, handle_abort=False)
+
+    if rc != 0:
+        msg = "%s: Reboot command failed" % task_action
+        raise _ReturnResultException(msg, rc=rc, stdout=stdout, stderr=stderr)
+
+
+def _reset_connection(task_action, connection, ignore_errors=False):  # type: (str, ConnectionBase, bool) -> None
+    """Resets the connection handling any errors"""
+    def _wrap_conn_err(func, *args, **kwargs):
+        try:
+            func(*args, **kwargs)
+
+        except (AnsibleError, RequestsConnectionError, RequestsTimeout) as e:
+            if ignore_errors:
+                return False
+
+            raise AnsibleError(to_native(e))
+
+        return True
+
+    # While reset() should probably better handle this some connection plugins don't clear the existing connection on
+    # reset() leaving resources still in use on the target (WSMan shells). Instead we try to manually close the
+    # connection then call reset. If it fails once we want to skip closing to avoid a perpetual loop and just hope
+    # reset() brings us back into a good state. If it's successful we still want to try it again.
+    if getattr(_reset_connection, 'do_close', True):
```

Comment:
Smuggling contextual ephemeral state in a module-global function object makes me nervous, especially for something where it might be possible in the not-too-distant future for more than one of these to be running in a process at once.

Without something like `contextvars`, it seems like passing an internal surrogate state object in for this might be the best option- stashing it on `connection` is equally dangerous, since it's not "your" object.